pkg/transport/process/util.go

```diff
@@ -0,0 +1,244 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+package process
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	cdv2 "github.com/gardener/component-spec/bindings-go/apis/v2"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// ComponentDescriptorFile is the filename of the component descriptor in a processor message tar archive
+	ComponentDescriptorFile = "component-descriptor.yaml"
+
+	// ResourceFile is the filename of the resource in a processor message tar archive
+	ResourceFile = "resource.yaml"
+
+	// ResourceBlobFile is the filename of the resource blob in a processor message tar archive
+	ResourceBlobFile = "resource-blob"
+)
+
+// WriteProcessorMessage writes a component descriptor, resource and resource blob as a processor
+// message (tar archive with fixed filenames for component descriptor, resource, and resource blob)
+// which can be consumed by processors.
+func WriteProcessorMessage(cd cdv2.ComponentDescriptor, res cdv2.Resource, resourceBlobReader io.Reader, w io.Writer) error {
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	marshaledCD, err := yaml.Marshal(cd)
+	if err != nil {
+		return fmt.Errorf("unable to marshal component descriptor: %w", err)
+	}
+
+	if err := writeFileToTARArchive(ComponentDescriptorFile, bytes.NewReader(marshaledCD), tw); err != nil {
+		return fmt.Errorf("unable to write %s: %w", ComponentDescriptorFile, err)
+	}
+
+	marshaledRes, err := yaml.Marshal(res)
+	if err != nil {
+		return fmt.Errorf("unable to marshal resource: %w", err)
+	}
+
+	if err := writeFileToTARArchive(ResourceFile, bytes.NewReader(marshaledRes), tw); err != nil {
+		return fmt.Errorf("unable to write %s: %w", ResourceFile, err)
+	}
+
+	if resourceBlobReader != nil {
+		if err := writeFileToTARArchive(ResourceBlobFile, resourceBlobReader, tw); err != nil {
+			return fmt.Errorf("unable to write %s: %w", ResourceBlobFile, err)
+		}
+	}
+
+	return nil
+}
+
+func writeFileToTARArchive(filename string, contentReader io.Reader, outArchive *tar.Writer) error {
+	tempfile, err := ioutil.TempFile("", "")
+	if err != nil {
+		return fmt.Errorf("unable to create tempfile: %w", err)
+	}
+	defer tempfile.Close()
+
+	if _, err := io.Copy(tempfile, contentReader); err != nil {
```

Comment:
this call should already return the size of the file. so no need for the fstat command.